packages/utils/src/lock.ts

```diff
@@ -0,0 +1,54 @@
+import { EventEmitter } from 'events'
+
+enum LockEvent {
+  Unlock = 'unlock',
+}
+
+// Lock which can be used to ensure mutual exclusion in concurrent code.
+export class Lock {
+  private locked: boolean = false
+  private emitter: EventEmitter
+
+  constructor() {
+    this.emitter = new EventEmitter()
+  }
+
+  // Attempt to aquire the lock without blocking.
+  //
+  // @returns {boolean} True if the lock was aquired.
+  tryAcquire(): boolean {
+    if (!this.locked) {
+      this.locked = true
+      return true
+    }
+    return false
```

Comment:
did you think about reentrancy behavior? right now it looks like this is non-reentrant but could be made reentrant here.

Reply:
Reentrancy would require something analogous to thread-local storage, which based on a quick search is implemented by `async_hooks` which is an experimental feature since Node 8. Given that it's not very simple to implement with the tools available, and there isn't a need for it at the moment, I would opt not to implement it right now.
https://nodejs.org/api/async_hooks.html

I did add a comment about it being non-reentrant though.

Reply:
>  and there isn't a need for it at the moment

why? are we assuming single threading...?

Reply:
It's not needed because we can ensure when we use the lock that `lock.acquire` is never called from the "thread" that holds the lock. (Although it's true that we only have one true thread in Node, so what the "thread" I am referring to here is an `async`/`await` chain of execution. I am not sure what the real name is for it.) (As a reference, reentrancy is not a required feature. Golang, for instance, [does not implement it](https://stackoverflow.com/questions/14670979/recursive-locking-in-go))